src/rust/lib/prelude/src/data/semigroup.rs

```diff
@@ -0,0 +1,158 @@
+//! In mathematics, a semigroup is an algebraic structure consisting of a set together with an
+//! associative binary operation. A semigroup generalizes a monoid in that there might not exist an
+//! identity element. It also (originally) generalized a group (a monoid with all inverses) to a
+//! type where every element did not have to have an inverse, thus the name semigroup.
+
+use std::collections::HashMap;
+use std::hash::BuildHasher;
+use std::hash::Hash;
+use std::iter::Extend;
+
+
+
+// =================
+// === Semigroup ===
+// =================
+
+/// Mutable Semigroup definition. Impls should satisfy the associativity law:
+/// `x.concat(y.concat(z)) = x.concat(y).concat(z)`, in symbolic form:
+/// `x <> (y <> z) = (x <> y) <> z`
+pub trait Semigroup : Clone {
+    /// An associative operation.
+    fn concat_mut(&mut self, other:&Self);
+
+    /// An associative operation which consumes the argument.
+    fn concat_mut_take(&mut self, other:Self) {
+        self.concat_mut(&other)
+    }
+
+    /// Repeat a value n times. Given that this works on a Semigroup it is allowed to fail if you
+    /// request 0 or fewer repetitions, and the default definition will do so.
+    fn stimes(&mut self, n:usize) {
+        let val = self.clone();
+        for _ in 0..n-1 {
+            self.concat_mut(&val)
+        }
+    }
+}
+
+/// Immutable Semigroup definition. Impls should satisfy the associativity law:
+/// `x.concat(y.concat(z)) = x.concat(y).concat(z)`, in symbolic form:
+/// `x <> (y <> z) = (x <> y) <> z`
+pub trait SemigroupIm : Clone {
+    /// An associative operation.
+    fn concat(&self, other:&Self) -> Self;
+
+    /// An associative operation which consumes the argument.
+    fn concat_take(&self, other:Self) -> Self {
+        self.concat(&other)
+    }
+
+    /// Repeat a value n times. Given that this works on a Semigroup it is allowed to fail if you
+    /// request 0 or fewer repetitions, and the default definition will do so.
+    fn stimes(&self, n:usize) -> Self where Self:Clone {
```

Comment:
why not just name this `times`? Are there any other traits that want to define their own `times`?

Reply:
`times` is implemented by `Monoid`. As agreed via phone, changing this name to `partial_times` instead.